src/AdvancedHMC.jl

```diff
@@ -146,7 +152,12 @@ function Hamiltonian(metric::AbstractMetric, ℓ; kwargs...)
     ℓπ = if cap === LogDensityProblems.LogDensityOrder{0}()
         # In this case ℓ does not support evaluation of the gradient of the log density function
         # We use ForwardDiff to compute the gradient
-        LogDensityProblemsAD.ADgradient(Val(:ForwardDiff), ℓ; kwargs...)
+        _logdensity = Base.Fix1(LogDensityProblems.logdensity, ℓ)
+        _logdensity_and_gradient = function (x)
+            prep = DI.prepare_gradient(_logdensity, adtype, x)
```

Comment:
Preparing outside the closure would be more efficient, but users have to be warned I guess

Reply:
I agree we should prepare outside the closure, but since we need the log density and its gradient of the model as a function in Hamiltonian, how can we prepare in advance?

Reply:
just move the `prep` one line up?

Reply:
Do you have access to a "typical input" somewhere?

Reply:
AFAIK, the `Hamiltonian` is basically used for keeping the log density and its gradient function, and "typical input" is used in the subsequent sampling process which may not be possible to pass in here.